runtime/compiler/trj9/runtime/J9ValueProfiler.hpp

```diff
@@ -28,266 +28,842 @@
 #include "env/TRMemory.hpp"  // for TR_Memory, etc
 #include "env/jittypes.h"    // for TR_ByteCodeInfo, etc
 #include "il/DataTypes.hpp"  // for DataTypes
-#include "infra/Monitor.hpp" // for TR::Monitor
+#include "infra/CriticalSection.hpp" // for OMR::CriticalSection
+#include "compile/Compilation.hpp"
+#include "infra/vector.hpp"         // for TR::vector
 
 extern TR::Monitor *vpMonitor;
 
-#define MAX_UNLOCKED_PROFILING_VALUES 5
+#ifdef TR_HOST_64BIT
+#define HIGH_ORDER_BIT (0x8000000000000000)
+#else
+#define HIGH_ORDER_BIT (0x80000000)
+#endif
+
+#define ARRAY_MAX_NUM_VALUES (5)
+#define LINKEDLIST_MAX_NUM_VALUES (20)
 
 class TR_ResolvedMethod;
-namespace TR { class Compilation; }
 template <class T> class List;
 template <class T> class ListElement;
 
+/**
+ * Profiling info consists of two class hierarchies: TR_AbstractInfo & TR_AbstractProfilerInfo
+ *
+ * TR_AbstractProfilerInfo is the persistent data structures use by the various profiling
+ * implementations. They can manage a variety of basic types, such as uint32_t and uint64_t.
+ * The class hierarchy, with T being uint32_t, uint64_t or TR_ByteInfo:
+ *
+ * TR_AbstractProfilerInfo
+ *    TR_LinkedListProfilerInfo<T>
+ *    TR_ArrayProfilerInfo<T>
+ *
+ * TR_AbstractInfo provides a common API for more complex data types, such as addresses and
+ * BigDecimals, consistent across the various sources of profiling information.
+ * The class hierarchy, with T being uint32_t, uint64_t or TR_ByteInfo:
+ *
+ * TR_AbstractInfo
+ *    TR_GenericValueInfo<T>
+ *       TR_AddressInfo
+ *       TR_BigDecimalInfo
+ */
+
+/**
+ * Kinds of value profiling information
+ */
+enum TR_ValueInfoKind
+   {
+   ValueInfo            = 0x00,
```

Comment:
If these are not being used as individual bits why do we need to assign numbers in a flags style? Can't we use the default numbering or standard decimal numbering so we don't think a bit or will do something useful which is the usual intent between a hex value assignment eg 0x1, 0x2, 0x4, 0x8 etc.

Reply:
Originally, these were assigned individual bits, however, that ultimately wasn't required. These have been corrected to use the default numbering.